planner/core/stats.go

```diff
@@ -146,9 +147,121 @@ func (ds *DataSource) DeriveStats(childStats []*property.StatsInfo) (*property.S
 			break
 		}
 	}
+	// consider the IndexMergePath. Now, we just generate `IndexMergePath` in DNF case.
+	if len(ds.pushedDownConds) > 0 && len(ds.possibleAccessPaths) > 1 && ds.ctx.GetSessionVars().EnableIndexMerge {
+		needConsiderIndexMerge := true
+		for i := 1; i < len(ds.possibleAccessPaths); i++ {
+			if len(ds.possibleAccessPaths[i].accessConds) != 0 {
+				needConsiderIndexMerge = false
+				break
+			}
+		}
+		if needConsiderIndexMerge {
+			ds.generateIndexMergeOrPaths()
+		}
+	}
 	return ds.stats, nil
 }
 
+// getIndexMergeOrPath generates all possible IndexMergeOrPaths.
+func (ds *DataSource) generateIndexMergeOrPaths() {
+	usedIndexCount := len(ds.possibleAccessPaths)
+	for i, cond := range ds.pushedDownConds {
+		sf, ok := cond.(*expression.ScalarFunction)
+		if !ok || sf.FuncName.L != ast.LogicOr {
+			continue
+		}
+		var partialPaths = make([]*accessPath, 0, usedIndexCount)
+		dnfItems := expression.FlattenDNFConditions(sf)
+		for _, item := range dnfItems {
+			cnfItems := expression.SplitCNFItems(item)
+			itemPaths := ds.accessPathsForConds(cnfItems, usedIndexCount)
+			if len(itemPaths) == 0 {
+				partialPaths = nil
+				break
+			}
+			partialPath := ds.buildIndexMergePartialPath(itemPaths)
+			if partialPath == nil {
+				partialPaths = nil
+				break
+			}
+			partialPaths = append(partialPaths, partialPath)
+		}
+		if len(partialPaths) > 1 {
+			possiblePath := ds.buildIndexMergeOrPath(partialPaths, i)
+			if possiblePath != nil {
+				ds.possibleAccessPaths = append(ds.possibleAccessPaths, possiblePath)
+			}
+		}
+	}
+}
+
+// accessPathsForConds generates all possible index paths for conditions.
+func (ds *DataSource) accessPathsForConds(conditions []expression.Expression, usedIndexCount int) []*accessPath {
+	var results = make([]*accessPath, 0, usedIndexCount)
+	for i := 0; i < usedIndexCount; i++ {
+		path := &accessPath{}
+		var noIntervalRanges bool
+		var err error
+		if ds.possibleAccessPaths[i].isTablePath {
+			path.isTablePath = true
+			noIntervalRanges, err = ds.deriveTablePathStats(path, conditions)
+		} else {
+			// now these conditions are too strict.
+			// for example, a sql `select * from t where a > 1 or (b < 2 and c > 3)` and table `t` with indexes on a and b separately.
+			// we can generate a `IndexMergePath` with table filter `a > 1 or (b < 2 and c > 3)`.
+			// TODO: solve the above case
+			path.index = ds.possibleAccessPaths[i].index
+			noIntervalRanges, err = ds.deriveIndexPathStats(path, conditions)
+		}
+		if err != nil {
+			logutil.BgLogger().Info("can not derive statistics of a path")
+		}
+		// if accessConds is empty or tableFilter is not empty, we ignore the access path.
+		if len(path.tableFilters) > 0 || len(path.accessConds) == 0 {
+			continue
+		}
+		// If we have point or empty range, just ignore other possible paths.
+		if noIntervalRanges || len(path.ranges) == 0 {
+			results = append(results, path)
```

Comment:
We need the unique check as well. For example, if the condition is `a = 1 and b = 2 and c > 1`, both `a = 1` and `b = 2` would return true for `noIntervalRanges`, while there may have 300 rows satisfying `a = 1` or `b = 1`, and only 1 row satisfying `c > 1`, so we can only prune others when the column is unique. You can see `DataSource::DeriveStats` for reference.

Also, if the check conditions are held, we need to empty the `results` as well before appending current path into it.